contracts/NodeVersionBeacon.cdc

```diff
@@ -85,337 +105,414 @@ pub contract NodeVersionBeacon {
-    /// maintained in this contract
-    pub resource NodeVersionAdmin {
-        /// Update the minimum version to take effect at the given block height
-        pub fun addVersionBoundaryToTable(targetBlockHeight: UInt64, newVersion: Semver) {
-            pre {
-                targetBlockHeight > getCurrentBlock().height + NodeVersionBeacon.versionUpdateBuffer
-                    : "Target block height occurs too soon to update EN version."
-            }
+            // No boundaries defined beyond current block, safe to make changes
+            if NodeVersionBeacon.firstUpcomingBoundary == nil {
+                NodeVersionBeacon.versionBoundaryFreezePeriod = newFreezePeriod
+                return
+            } 
+            
+            let nextBlockBoundary = NodeVersionBeacon.versionBoundaryBlockList[NodeVersionBeacon.firstUpcomingBoundary!]
 
-            // Insert mapping into version table
-            NodeVersionBeacon.versionTable.insert(
-                key: targetBlockHeight,
-                newVersion
+            // Ensure that the we're not currently within the old or new freeze period
+            // of the next block height boundary
+            assert(
+                currentBlockHeight + NodeVersionBeacon.versionBoundaryFreezePeriod < nextBlockBoundary &&
+                currentBlockHeight + newFreezePeriod < nextBlockBoundary,
+                message: "Updating buffer now breaks version boundary update expectations. Try updating buffer after next version boundary."
             )
-            // Insert the version boundary's target block height into the
-            // array maintaining all upcoming block height boundaries
-            NodeVersionBeacon.insertUpcomingBlockBoundary(targetBlockHeight)
 
-            // Set the flag to true so the event will be emitted next time emitChanges is called
-            NodeVersionBeacon.tableUpdated = true
+            NodeVersionBeacon.versionBoundaryFreezePeriod = newFreezePeriod
+            
+            emit NodeVersionBoundaryFreezePeriodChanged(freezePeriod: newFreezePeriod)
         }
+    }
 
-        /// Deletes the last entry in versionTable which defines an upcoming version boundary
-        /// Could be helpful in case rollback is needed
-        pub fun deleteUpcomingVersionBoundary(blockHeight: UInt64): Semver {
-            pre {
-                NodeVersionBeacon.versionTable.keys.contains(blockHeight): "No boundary defined at that blockHeight."
-                NodeVersionBeacon.upcomingBlockBoundaries.contains(blockHeight): "Boundary not defined in upcomingBlockBoundaries."
+    /// Heartbeat resource that emits the version beacon event and keeps track of upcoming versions.
+    /// This resource should always be held only by the service account,
+    /// because the service account should be the only one emitting the event, 
+    /// and only during the system transaction
+    pub resource Heartbeat {
```

Comment:
Is it a problem that we also use `Heartbeat` as the name for the epoch resource that advances epoch phases? Might be a bit confusing?

I guess they both effectively are doing some action every block so it might make sense that they have the same names

Reply:
Exactly, both functions have the same use: they need to be called during the system transaction every block. So I thought it might be better that they have the same name.